### PR TITLE
fix: gracefully quit panicking Cmd

### PR DESCRIPTION
**Pull request by joerdav**
Fixes: #234

### CODE REVIEW

**Comment by muesli** [comment]
We should try and add a test for this.

**Comment by blvrd** [comment]
I'd be happy to help get this over the finish line if there's anything else that needs to be done. It would be nice to not need to `reset` the terminal if there's a panic inside a `tea.Cmd`.

**Comment by meowgorithm** [comment]
@blvrd can you verify that it fixes #1127?

**Comment by blvrd** [comment]
No, it doesn't fix #1127 since the panic in the reproduction script is happening outside of Bubbletea. If the `recoverFromPanic` method from this PR was exported, you _could_ fix #1127 with this approach:

```go
package main

import (
	"fmt"
	"os"
	"time"

	tea "github.com/charmbracelet/bubbletea"
)

type model struct{}

func (m model) Init() tea.Cmd {
	return nil
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	return m, nil
}

func (m model) View() string {
	return "I will panic"
}

func main() {
	m := model{}
        p := tea.NewProgram(m)

	go func() {
                defer p.RecoverFromPanic() // recover here
		time.Sleep(2 * time.Second)
		panic("")
	}()

	if _, err := p.Run(); err != nil {
		fmt.Println("Uh oh, we encountered an error:", err)
		os.Exit(1)
	}
}
```

This doesn't feel like a great solution though. Maybe you'd just have to be responsible for your own panic recovery for goroutines that execute outside of the scope of the `tea.Program`?

File: tea.go [hunk 1]
```diff
@@ -300,9 +304,22 @@ func (p *Program) handleCommands(cmds chan Cmd) chan struct{} {
 	return ch
 }
 
+func (p *Program) recoverFromPanic() {
+	if r := recover(); r != nil {
+		fmt.Printf("Caught panic:\n\n%s\n\nRestoring terminal...\n\n", r)
+		debug.PrintStack()
+		p.errs <- fmt.Errorf("%v", r)
+		return
+	}
```
Comment (meowgorithm) [inline tea.go:313]: I believe this is very close to code used in `Program.Run` so we should probably use this method there as well. That said, this method appears to be missing the `p.shutdown(true)` we see in `Program.Run`: is that intentional?

Comment (blvrd) [reply tea.go:313]: cc: @joerdav

**Comment by meowgorithm** [comment]
Just a heads up that we're fixing this over at #1134; it's nuanced and we want to be very deliberate with this one. @joerdav, we appreciate you flagging this one!

**Comment by meowgorithm** [comment]
Hello again! Just a note that the fix for this is now available in Bubble Tea v1.1.1.

**Pull request closed without merging**